Redesign Trace Timeline

tacigar opened the pull request:
resolve #2672 
resolve #2525 
resolve #2195

tacigar committed:
Add TraceTimeline

tacigar committed:
Update css

tacigar committed:
Set offset for tree rendering

tacigar committed:
Add TraceTree

tacigar commented:
### Current state

* Use SVG for trace graph
* Fix page header
* Add trace tree lines

<img width="1440" alt="Screenshot 2019-08-05 13 26 19" src="https://user-images.githubusercontent.com/19551419/62438936-aeef5f80-b784-11e9-8a58-e94ccb882623.png">

tacigar committed:
Add SpanDetail

tacigar committed:
Enable to click span row

tacigar commented:
### Current State

* Add span detail on the right side of the page.

In #2525 , I said that I wanted to set the span details on the right side of the page.
And I got some feedback about it.
As a result of thinking, by providing a button to collapse the span details, now I think that the point of concerns may be resolved.

<img width="1440" alt="Screenshot 2019-08-05 15 26 26" src="https://user-images.githubusercontent.com/19551419/62443328-99cefc80-b795-11e9-9234-9a076110966c.png">

tacigar committed:
Add expand button

tacigar committed:
Add expand buttons

tacigar committed:
Implement expand button

tacigar committed:
Fix css bug

tacigar committed:
Add TraceTimelineHeader

tacigar commented:
#### Current State

* Add TimeMarker
* Add expand/collapse button
* Add dummy buttons

<img width="1440" alt="Screenshot 2019-08-06 19 08 38" src="https://user-images.githubusercontent.com/19551419/62531941-66b46800-b87e-11e9-988c-db3b863d2eb2.png">

tacigar committed:
Add duration info to TraceTimeline

devinsba commented:
I like it, I'd love to see what it looks like with some of the more extreme example data. Like the depth > 10 data or 1000+ spans

codefromthecrypt commented:
Thanks for starting this effort. it is daring work!

I have an image with feedback. I think this is also a good time to add "site tags" at a stable place in the view as discussed before, though not necessarily in this change.

<img width="816" alt="Screenshot 2019-08-07 at 9 45 36 AM" src="https://user-images.githubusercontent.com/64215/62584895-901ad580-b90a-11e9-9faf-cdfb7604fa33.png">

cc @openzipkin/ui

tacigar committed:
Add MiniMap

tacigar committed:
Update SpanAnnotation design

tacigar committed:
Improve UX of Annotation

tacigar committed:
makeStyles -> withStyles

tacigar committed:
Add unit tests for SpanAnnotation

tacigar committed:
Add testing-library/jest-dom/extend-expect

tacigar committed:
Add unit tests for SpanAnnotations

tacigar committed:
Add unit tests for SpanTags

tacigar committed:
Update unit tests for SpanAnnotation

tacigar committed:
Add unit tests for SpanDetail

tacigar committed:
Unuse TraceMiniMap

tacigar committed:
makeStyles -> withStyles

tacigar committed:
Refactor TimeMarker

tacigar committed:
Add unit tests for TimeMarker

tacigar committed:
Add unit tests for TraceTimelineHeader

tacigar committed:
Refactor and Simplify TraceTimeline

tacigar committed:
Remove unused code

anuraaga commented:
I haven't been looking at lens prs much but noticed this commit

makeStyles -> withStyles

Any reason to go from hooks back to HoC? I've been migrating in the opposite direction so just curious :)

thien-do commented on zipkin-lens/src/components/TracePage/SpanDetail/SpanAnnotation.jsx:
```diff
@@ -0,0 +1,81 @@
+                // NOTE: a.timestamp % 1000 would save a string conversion but drops
+                // leading zeros.
+                value: moment(annotation.timestamp / 1000).format('MM/DD HH:mm:ss.SSS')
+                  + annotation.timestamp.toString().slice(-3),
```
I guess the result of this would be something like `08/18 16:13:02.123456`? Since the fractional part of the second (`123456`) is longer than 3, I think we can do a small touch and use `_` to improve readability (so it would be `02.123_456`).

I believe it's common in several programming languages and is coming to JS too: https://github.com/tc39/proposal-numeric-separator#references

thien-do replied:
Also I think it would be useful to extract this functionalitity to something like a global util of this package? (especially for its comment)

tacigar replied:
Thank you for your advice!
Done 🙇

thien-do commented on zipkin-lens/src/components/TracePage/SpanDetail/SpanAnnotationGraph.jsx:
```diff
@@ -0,0 +1,171 @@
+import classnames from 'classnames';
+
+import { selectServiceColor } from '../../../colors';
+import { spanAnnotationsPropTypes } from '../../../prop-types';
```
should we use absolute import here? Or do we need to wait for #2410 first?

thien-do commented on zipkin-lens/src/components/TracePage/SpanDetail/SpanDetail.jsx:
```diff
@@ -0,0 +1,78 @@
+    width="100%"
+    minHeight={minHeight}
+    borderLeft={1}
+    borderColor="grey.300"
```
oh we can't use `@material-ui/core/colors/grey` here? It seems safer than a string

thien-do commented on zipkin-lens/src/components/TracePage/SpanDetail/SpanTags.jsx:
```diff
@@ -0,0 +1,71 @@
+    fontWeight: 'bold',
+  },
+  value: {
+    fontSize: '1.05rem',
```
Is this a fractional value?

In general I think using px is better than rem since it's difficult to know what's the final value in case of rem.

The reason some prefer rem is "to make it scale with users' setting", however, in practice, users are most likely to zoom their browser (e.g. `cmd +`) instead of going to settings and set a different font size. In these cases, px works much better

thien-do commented on zipkin-lens/src/components/TracePage/SpanDetail/SpanTags.jsx:
```diff
@@ -0,0 +1,71 @@
+ */
+import PropTypes from 'prop-types';
+import React from 'react';
+import { withStyles } from '@material-ui/styles';
```
Why do we prefer HOC API overs styled-component API or hook API?

tacigar replied:
When using the Hooks API, I need to mock makeStyles.
like https://github.com/mui-org/material-ui/issues/15404#issuecomment-497373555
This is why I chose the HOC style...

thien-do commented on zipkin-lens/src/components/TracePage/TraceMiniMap/TraceMiniMap.jsx:
```diff
@@ -0,0 +1,65 @@
+}) => (
+  <Box
+    width="100%"
+    height="70px"
```
I thought we are using a 8px grid (in material ui)?

thien-do commented on zipkin-lens/src/components/TracePage/TraceMiniMap/TraceMiniMapGraph.jsx:
```diff
@@ -0,0 +1,51 @@
+  spans: detailedSpansPropTypes.isRequired,
+};
+
+const graphHeight = 55; // px
```
Why not 56? Is this 56 - 1px?

thien-do commented on zipkin-lens/src/components/TracePage/TraceSummary.jsx:
```diff
@@ -11,46 +11,78 @@
+import React, { useState, useCallback, useMemo } from 'react';
 import Box from '@material-ui/core/Box';
-// import { AutoSizer } from 'react-virtualized';
+import { AutoSizer } from 'react-virtualized';
```
just a minor thing :D I think we can upgrade to react-window soon!

tacigar replied:
Thank you for your information!

thien-do commented on zipkin-lens/src/components/TracePage/TraceTimeline/TraceTree.jsx:
```diff
@@ -0,0 +1,270 @@
+  closedSpans,
+  onSpanToggleButtonClick,
+  classes,
+}) => {
```
At a glance it feels like this component is doing quite many things here.. but can't be sure until measure it :D

thien-do replied:
But I do think these logics should be extract to out of this function/component. They are not really simple.

imballinst replied:
How often this component will re-render? I think [useMemo](https://reactjs.org/docs/hooks-reference.html#usememo) can be quite useful here to increase rendering performance (just in case if it is slow -- otherwise we will be only doing premature optimization)

thien-do commented on zipkin-lens/src/components/TracePage/TraceTimelineHeader/TimeMarker.jsx:
```diff
@@ -0,0 +1,96 @@
+    position: 'absolute',
+  },
+  'label--last': {
+    left: 'initial',
```
oh this is interesting. I often see `auto` at most

codefromthecrypt commented:
@anuraaga @dvkndn very appreciate you having a look! It helps @tacigar feel less alone (also good for the code) :D

tacigar committed:
Add key props to elements of TraceTree

tacigar committed:
Change timestamp format of annotation

tacigar committed:
Refactor SpanDetail

tacigar committed:
Add some comments

tacigar committed:
Fix SpanAnnotation bug

tacigar committed:
Remove TraceMiniMap

tacigar committed:
Refactor TraceTree a little

imballinst commented on zipkin-lens/src/components/TracePage/SpanDetail/SpanAnnotation.jsx:
```diff
@@ -0,0 +1,81 @@
+
+const style = {
+  cell: {
+    paddingTop: '8px',
```
Hi! Curious, isn't it better to use `theme.spacing(1)` in this case? Or anything that is equivalent to `8px` if this UI is using custom Theme.

Makes it easier for global customization as well _if_ there needs to be a global padding change in the future.

imballinst commented on zipkin-lens/src/components/TracePage/SpanDetail/SpanDetail.jsx:
```diff
@@ -0,0 +1,78 @@
+
+const style = theme => ({
+  root: {
+    backgroundColor: grey[100],
```
Would it be better if we use `theme.palette.grey[100]` instead?

imballinst commented on zipkin-lens/src/components/TracePage/TraceSummary.jsx:
```diff
@@ -11,46 +11,78 @@
+      [spanId]: true,
     });
-  }, []);
+  }, [closedSpans]);
```
Idea: instead of having `[closedSpans]` as deps, why don't use the pattern as follows:

```js
setClosedSpans(oldSpans => {
  return { 
    ...oldSpans,
    [spanId]: oldSpans[spanId] ? undefined : true
  };
});
```
Otherwise, each time we set new value to `closedSpans`, the value of `handleSpanToggleButtonClick` will change.

tacigar replied:
Oh! Thank you for your advice!
I'll do so.

imballinst reviewed (commented):
Hi! Left some reviews -- hopefully it helps.

I second what everyone else said about usage of `withStyles` instead of `makeStyles`. On top of that, I _think_ some of the colors should be using the values from the `theme` instead of imported directly from `@material-ui/core/colors/*` package (although they yield the same thing). Same thing for stuff such as `padding`.

tacigar commented:
Thank you everyone for your review!
I am very sorry my work is late.
I'm busy with other work right now...
I will resume this work as soon as possible :bowing_man:

tacigar committed:
Use theme.spacing for table cell padding

tacigar committed:
Update handleSpanToggleButtonClick

tacigar committed:
Refactor TraceTree

tacigar commented:
I'll implement re-root feature.

codefromthecrypt commented:
@tacigar  🙇 when you are ready for review, ask me (🙇 optional when asking :P)

tacigar commented:
@adriancole ok and sorry for slow works...

codefromthecrypt commented:
I have made much slower work.. some never finished. no worries @tacigar

tacigar committed:
Reroot trace tree when a span is clicked twice

tacigar committed:
Disable reroot feature if the trace does not have a root span

tacigar committed:
Refactor TraceSummary and TraceTimeline

tacigar committed:
Implement reset-reroot button

tacigar committed:
Change background color of the span row which is focused

tacigar committed:
Zoom up when rerooted

tacigar committed:
Set minimum valud for span's bar width

tacigar committed:
Give default value span bar's width

tacigar committed:
Add right margin for timeline

tacigar commented:
![reroot](https://user-images.githubusercontent.com/19551419/65675614-733c8c00-e089-11e9-8598-0dc5bd5b21f9.gif)

Reroot feature has been almost done.
But I have found some bugs in the side window which shows annotation and tag info, so I have to fix it.

codefromthecrypt commented:
Thanks for the update igarashi!

tacigar committed:
Update comments

tacigar committed:
Fix bugs of reroot feature

Fix a strange behavior when a span has no duration and only that span is displayed in the trace timeline.

tacigar committed:
Open and close span detail

tacigar committed:
Implement Expand/Collapse buttons

tacigar committed:
Fix detailed span prop type

tacigar committed:
Fix a bug related to span with the same annotation value

tacigar committed:
Fix bugs of duplicated key of annotations

tacigar committed:
Fix bugs of duplicated key of tags

tacigar committed:
Fix unit test a bit

tacigar commented:
@adriancole 
Almost done.
Could you please try to use it?

codefromthecrypt commented:
yes!

codefromthecrypt commented:
noticed some shakes in the list view...

![Shake, shake, shake, Senora](https://user-images.githubusercontent.com/64215/65752173-cc1a2c00-e13e-11e9-8c06-fc98f2eb4115.gif)

codefromthecrypt commented:
I find scrolling in the main screen on the large trace (smartthings-mobile-web-install.json from @beckje01 @llinder) kindof hard to follow. It starts becoming easier when I root into one. A small glitch is that the up control bombs.

![up bomb](https://user-images.githubusercontent.com/64215/65752865-4d25f300-e140-11e9-9c71-1a8a83ad9e66.gif)

I can see the hierarchy easier when looking at a long trace in the former design vs this one at the moment

![old](https://user-images.githubusercontent.com/64215/65753105-d89f8400-e140-11e9-90ec-eedb5b3e1450.gif)

codefromthecrypt commented:
my guess is some more horizontal space could be used to represent the span durations. this should restore some vague idea of the hierarchy without relying completely on the light lines (I do like the lines)

tacigar commented:
@adriancole 
> my guess is some more horizontal space could be used to represent the span durations.

excuse me.
specifically, where should I use more horizontal space? 😃

tacigar commented:
> noticed some shakes in the list view...

hmm..
I cannot reenact that phenomenon.

anuraaga commented on zipkin-lens/src/components/DiscoverPage/DiscoverPage.jsx:
```diff
@@ -266,37 +264,47 @@ const DiscoverPage = ({ history, location }) => {
-          <TraceJsonUploader />
-          <TraceIdSearchInput />
+      <Box pl={3} pr={3}>
+        <Box width="100%" display="flex" justifyContent="space-between">
```
First time looking at lens code :)

Out of curiosity, have you thought about using `Grid`? Maybe it makes this sort of layout simpler.

anuraaga commented on zipkin-lens/src/components/TracePage/SpanDetail/SpanAnnotation.jsx:
```diff
@@ -0,0 +1,71 @@
+    </Box>
+    <Paper>
+      <Table>
+        <TableBody data-testid="span-annotation--table-body">
```
Probably not this PR, but it'd be good to strip these with https://www.npmjs.com/package/babel-plugin-react-remove-properties

zipkin UI has tons of DOM, so these extra attributes look like they can have significant enough impact on performance compared to other apps.

anuraaga commented on zipkin-lens/src/components/TracePage/SpanDetail/SpanAnnotationGraph.jsx:
```diff
@@ -0,0 +1,187 @@
+    minTs = annotations[0].timestamp;
+    maxTs = minTs + 1;
+  } else {
+    minTs = _.minBy(annotations, 'timestamp').timestamp;
```
Probabyl better to import the functions from lodash instead of all of lodash.

anuraaga commented on zipkin-lens/src/components/TracePage/SpanDetail/SpanAnnotation.jsx:
```diff
@@ -0,0 +1,71 @@
+
+SpanAnnotation.propTypes = propTypes;
+
+export default withStyles(style)(SpanAnnotation);
```
Also how about adding `React.memo` to all the components? Avoiding rerendering is very important with all this DOM.

tacigar replied:
I want to think a little more about whether to use `React.memo` for all the components.
But using `React.memo` on the trace timeline clearly improved rendering performance.

**Before**

![before-optimization](https://user-images.githubusercontent.com/19551419/65944886-a4dc9b00-e46d-11e9-99d5-1ea8935418f1.gif)

**After**

![after-optimization](https://user-images.githubusercontent.com/19551419/65944905-b02fc680-e46d-11e9-8d63-eeb30f60f69b.gif)

Thank you for your advice!

anuraaga replied:
That's a nice improvement! I'm guessing there might be other good places like each individual trace (make sure not to rerender when expanding a trace) - but often it's just simplest to reason about to use `memo` as much as possible since

- We have so many components anyways, adding the HoC isn't a significant impact
- Either a component rerenders, meaning the computation of memo was a bit of a waste, or it doesn't meaning it was worth it. But even for very simple components the rerender (virtual DOM operations) should always be much more expensive then computing the memo.

But will leave it to you how much you add it :)

anuraaga commented on zipkin-lens/src/components/TracePage/SpanDetail/SpanAnnotations.jsx:
```diff
@@ -0,0 +1,100 @@
+    setAreAllAnnotationsOpened(false);
+  }, [span.spanId]);
+
+  const currentAnnotation = useMemo(() => span.annotations.find((annotation) => {
```
This is `true` or `false`, so a name like `isCurrentAnnotation` or `currentAnnotationOpened` would be easier to follow.

tacigar replied:
Oh, `currentAnnotation` is not boolean value.

anuraaga replied:
Ah oops, that was just the argument to `find` - got confused

anuraaga commented on zipkin-lens/src/components/TracePage/SpanDetail/SpanAnnotations.jsx:
```diff
@@ -0,0 +1,100 @@
+  }, [areAllAnnotationsOpened]);
+
+  return (
+    <React.Fragment>
```
`<>`?

tacigar committed:
Fix up-control bombs bug

anuraaga commented on zipkin-lens/src/components/TracePage/SpanDetail/SpanDetail.jsx:
```diff
@@ -0,0 +1,74 @@
+      <SpanAnnotations span={span} />
+    </Box>
+    <Box pt={1} pl={2} pr={2} pb={1.5}>
+      <Box fontWeight="bold" fontSize="1.4rem" mb={0.5}>
```
Can we get font settings from the theme?

codefromthecrypt commented:
@tacigar sorry about not being clear about horizontal. I will re-pull your change and take a screen shot 🙇

tacigar commented:
thank you!

anuraaga commented on zipkin-lens/src/util/timestamp.js:
```diff
@@ -11,15 +11,26 @@
+// with just milliseconds and then append the last 3 digits of the timestamp
+// which are the microseconds.
+// NOTE: a.timestamp % 1000 would save a string conversion but drops leading zeros.
+export const formatTimestampMillis = timestamp => `${formatTimestamp(timestamp)}_${timestamp.toString().slice(-3)}`;
```
Isn't this `formatTimestampMicros`?

codefromthecrypt commented:
<img width="1024" alt="Screenshot 2019-09-30 at 2 15 39 PM" src="https://user-images.githubusercontent.com/64215/65854022-ed248c00-e38d-11e9-9b0b-2d404089427b.png">

codefromthecrypt commented:
PS no more bouncy results. thanks for fixing it.

tacigar committed:
Don't change the span bar width when collapse/expand spans

tacigar commented:
Those bars of new design are bars for showing span timestamp and duration, not for showing trace hierarchy...

<img width="1200" alt="スクリーンショット 2019-09-30 16 18 41" src="https://user-images.githubusercontent.com/19551419/65857274-7643bf00-e39e-11e9-84c8-7dd6b0908ac3.png">

But I can widen the width of the Hierarchy line tree like:

**before**

<img width="2032" alt="スクリーンショット 2019-09-30 16 24 40" src="https://user-images.githubusercontent.com/19551419/65857464-e2262780-e39e-11e9-9457-784bce49455a.png">


**after**

<img width="2032" alt="スクリーンショット 2019-09-30 16 24 17" src="https://user-images.githubusercontent.com/19551419/65857476-e6eadb80-e39e-11e9-8644-31e330be367e.png">

WDYT?

codefromthecrypt commented:
<img width="2032" alt="65857476-e6eadb80-e39e-11e9-8644-31e330be367e" src="https://user-images.githubusercontent.com/64215/65859329-7f329180-e39a-11e9-8335-52b6d3618c50.png">

tacigar committed:
Avoid re-rendering trace timeline if not needed

tacigar committed:
<React.Fragment> -> <>

tacigar committed:
Rename formatTimestampMillis to formatTimestampMicros

tacigar committed:
TMP

tacigar committed:
Implement service name badges

tacigar committed:
Implement bar

tacigar committed:
Change text position

tacigar committed:
Fix service name badge position

tacigar committed:
Fix timeMarker's position

tacigar committed:
Fix disabled collapse/expand button bugs

tacigar committed:
Refactor

tacigar commented:
@adriancole 
I've (re-)redesigned trace timeline.
Please try to use it 🙇 

<img width="1200" alt="スクリーンショット 2019-10-03 15 37 25" src="https://user-images.githubusercontent.com/19551419/66104511-bea7ea80-e5f3-11e9-960e-26bfe424565b.png">

tacigar committed:
Remove unnecessary lines

codefromthecrypt commented:
I am rushing to airport.. @zeagord @llinder @jorgheymans @jeqo maybe one of you can try this?

codefromthecrypt commented:
PS looks very nice!

codefromthecrypt commented:
THIS IS AWESOME

looking at @narayaruna trace, it is really easy to navigate. only small thing is I'd recommend not displaying [reset reroot] at all when a trace isn't re-rooted, because in this case it is distracting as the trace isn't big enough to need re-rooting anyway
![netflix](https://user-images.githubusercontent.com/64215/66174940-0ed18c00-e68a-11e9-898e-7b59661e8817.gif)

codefromthecrypt commented:
For @beckje01 @llinder really long trace, I think it "re-root" is more useful. double-click re-roots..

![oauth](https://user-images.githubusercontent.com/64215/66175135-09287600-e68b-11e9-8871-5e96c3b735b8.gif)

two suggestions:
* re-root should automatically scroll to the top of the new root
* double-click should also "un-re-root" It is intuitive to me at least.

tacigar committed:
Partial import lodash

tacigar committed:
Open span detail when span clicked

tacigar committed:
Don't show reroot button if the trace is not rerooted

tacigar committed:
Unreroot when the current root span is doulbe-clicked

codefromthecrypt commented:
pre-emptive suggestion if possible. when double click to un reroot. position the view on the span you un rerooted

tacigar committed:
Add key props

tacigar committed:
Swap + and -

tacigar commented:
>  position the view on the span you un rerooted

hmm..
It seems pretty difficult to implement it.
It may take some time, so I want to make another PR for that.

tacigar committed:
Remove unused files

tacigar committed:
Add timemarker to timeline

codefromthecrypt commented:
sounds good. let me know when you feel this is merge

tacigar committed:
Remove opacity

tacigar committed:
Merge package-lock.json

tacigar committed:
Don't write colors directly on DOM elements

anuraaga commented on zipkin-lens/src/components/DiscoverPage/DiscoverPage.jsx:
```diff
@@ -265,39 +264,49 @@ const DiscoverPage = ({ history, location }) => {
+                <Tab label="Dependencies" className={classes.tab} />
+              </Tabs>
+            </AppBar>
+            { /* 2rem is the height of the Appbar. */}
```
Can we extract a constant for the height of app bar and then we don't even need a comment?

anuraaga commented on zipkin-lens/src/components/TracePage/TraceSummary.jsx:
```diff
@@ -11,72 +11,183 @@
+  const handleChildrenToggle = useCallback((spanId) => {
+    setChildrenHiddenSpanIds(prevChildrenHiddenSpanIds => ({
+      ...prevChildrenHiddenSpanIds,
+      [spanId]: prevChildrenHiddenSpanIds[spanId] ? undefined : true,
```
Maybe stick with `false` instead of `undefined` for readability - setting to `undefined` doesn't remove from the map so there isn't any performance reason to do it and it looks a little odd next to `true`

anuraaga commented on zipkin-lens/src/components/TracePage/TraceSummary.jsx:
```diff
@@ -11,72 +11,183 @@
+    let max = 0;
+    rerootedTree.forEach((span) => {
+      let ts;
+      if (!span.duration) {
```
Just me, but I find this a bit more readable / less complexity as

```ts
let ts = span.timestamp;
if (span.duration) {
  ts += span.duration;
}
```

anuraaga replied:
Though incidentally, is it possible for `duration` to be undefined (as opposed to always being returned as 0 by API)?

codefromthecrypt replied:
in our model we have duration only valid with minimum value 1 for reasons like this :)

tacigar replied:
At least I found spans that don't have a duration in teatdata.

https://github.com/openzipkin/zipkin/blob/2c353e17c0a2fa8440fe01ba924aff36dd416aa3/zipkin-lens/testdata/netflix.json#L2-L41

anuraaga commented on zipkin-lens/src/components/TracePage/TraceSummary.jsx:
```diff
@@ -11,72 +11,183 @@
+  const startTs = useMemo(() => minBy(rerootedTree, 'timestamp').timestamp, [rerootedTree]);
+  const endTs = useMemo(() => {
+    let max = 0;
+    rerootedTree.forEach((span) => {
```
You might consider this

```ts
return rerootedTree.map((span) => {
  let ts = span.timestamp;
  if (span.duration) {
    ts += span.duration;
  }
  return ts;
}).reduce((a, b) => Math.max(a, b));
```

tacigar replied:
wow...
very high readability...

anuraaga commented on zipkin-lens/src/components/TracePage/TraceSummary.jsx:
```diff
@@ -11,72 +11,183 @@
+    const expandedSpanIds = shownTree
+      .filter(span => childrenHiddenSpanIds[span.spanId])
+      .reduce((acc, cur) => {
+        acc[cur.spanId] = undefined;
```
If doing the above comment, this also can be `false`

anuraaga commented on zipkin-lens/src/components/TracePage/TraceSummary.jsx:
```diff
@@ -11,72 +11,183 @@
 
+  const handleExpandButtonClick = useCallback(() => {
+    const expandedSpanIds = shownTree
+      .filter(span => childrenHiddenSpanIds[span.spanId])
```
Think it's good to explicitly convert to boolean when filtering `!!childrenHiddenSpanIds[span.spanId]`

codefromthecrypt commented:
Latest commit looks good. double-click action is nice!

codefromthecrypt commented:
somethings noticed:
TraceId: part is not completely precise when re-rooting. I think we should display the span ID when navigated here.. even better if the url could.

ex show "80f198ee56343ba864fe8b2a57d3eff7-e457b5a2e4d86bd1" if rooted to the ID e457b5a2e4d86bd1 in the trace 80f198ee56343ba864fe8b2a57d3eff7

nice plus would be to instead of showing

http://localhost:9411/zipkin/traces/80f198ee56343ba864fe8b2a57d3eff7
to show

http://localhost:9411/zipkin/traces/80f198ee56343ba864fe8b2a57d3eff7-e457b5a2e4d86bd1

I'm using hyphen format as that is the same as b3 single format https://github.com/openzipkin/b3-propagation#single-header

codefromthecrypt commented:
yes no duration is same as zero duration. valid duration starts at 1us. in
other words round up. this should mean truthy comparison is ok.

On Fri, Oct 4, 2019, 4:25 PM tacigar <notifications@github.com> wrote:

> *@tacigar* commented on this pull request.
> ------------------------------
>
> In zipkin-lens/src/components/TracePage/TraceSummary.jsx
> <https://github.com/openzipkin/zipkin/pull/2736#discussion_r331390720>:
>
> > +      if (allHiddenSpanIds[span.spanId] && span.childIds) {
> +        span.childIds.forEach((childId) => {
> +          allHiddenSpanIds[childId] = true;
> +        });
> +      }
> +    });
> +    return rerootedTree.filter(span => !allHiddenSpanIds[span.spanId]);
> +  }, [rerootedTree, childrenHiddenSpanIds]);
> +
> +  // Find the minumum and maximum timestamps in the shown spans.
> +  const startTs = useMemo(() => minBy(rerootedTree, 'timestamp').timestamp, [rerootedTree]);
> +  const endTs = useMemo(() => {
> +    let max = 0;
> +    rerootedTree.forEach((span) => {
> +      let ts;
> +      if (!span.duration) {
>
> At least I found spans that don't have a duration in teatdata.
>
>
> https://github.com/openzipkin/zipkin/blob/2c353e17c0a2fa8440fe01ba924aff36dd416aa3/zipkin-lens/testdata/netflix.json#L2-L41
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openzipkin/zipkin/pull/2736?email_source=notifications&email_token=AAAPVV6LGZA6R3YWRRVHQJLQM34YTA5CNFSM4IJHKXW2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCG4NXOY#discussion_r331390720>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAAPVVZK4CYTEEUX2OXZNM3QM34YTANCNFSM4IJHKXWQ>
> .
>

codefromthecrypt commented:
on duration being only valid starting at 1:
https://github.com/openzipkin/zipkin-api/blob/master/zipkin2-api.yaml#L464

you can make a comment like this

anuraaga commented:
If the API leaves out duration in some cases, it would be nice to set a duration of `1` when fetching and duration is missing since that is the restraint set by the model, and then we don't need to have as many checks at business logic call sites.

codefromthecrypt commented:
I dont quite understand what problem we are solving here. duration query is
expensive so we certainly should not add a constraint like that by default
especially since the query is trace scoped not span scoped. a 1us query
will match even a trace that has spans minus a duration.


also note that Netflix span here was a converted one from a different
system. it is not representative of a proper span, but a flushed span can
also be missing a duration.

On Fri, Oct 4, 2019, 4:58 PM Anuraag Agrawal <notifications@github.com>
wrote:

> If the API leaves out duration in some cases, it would be nice to set a
> duration of 1 when fetching and duration is missing since that is the
> restraint set by the model, and then we don't need to have as many checks
> at business logic call sites.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openzipkin/zipkin/pull/2736?email_source=notifications&email_token=AAAPVV2IUJUJP7R5CP5E2LDQM4AUFA5CNFSM4IJHKXW2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEALAKJY#issuecomment-538314023>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAAPVV663IKYDARR6NBYCU3QM4AUFANCNFSM4IJHKXWQ>
> .
>

anuraaga commented:
I didn't mean changing the query - just if the JavaScript encounters a span
with no duration it can fill in 1. The problem that is solved is that
instead of every call site having to check for empty duration, only one
does. It's not expensive.

On Fri, Oct 4, 2019, 19:03 Adrian Cole <notifications@github.com> wrote:

> I dont quite understand what problem we are solving here. duration query is
> expensive so we certainly should not add a constraint like that by default
> especially since the query is trace scoped not span scoped. a 1us query
> will match even a trace that has spans minus a duration.
>
>
> also note that Netflix span here was a converted one from a different
> system. it is not representative of a proper span, but a flushed span can
> also be missing a duration.
>
> On Fri, Oct 4, 2019, 4:58 PM Anuraag Agrawal <notifications@github.com>
> wrote:
>
> > If the API leaves out duration in some cases, it would be nice to set a
> > duration of 1 when fetching and duration is missing since that is the
> > restraint set by the model, and then we don't need to have as many checks
> > at business logic call sites.
> >
> > —
> > You are receiving this because you were mentioned.
> > Reply to this email directly, view it on GitHub
> > <
> https://github.com/openzipkin/zipkin/pull/2736?email_source=notifications&email_token=AAAPVV2IUJUJP7R5CP5E2LDQM4AUFA5CNFSM4IJHKXW2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEALAKJY#issuecomment-538314023
> >,
> > or mute the thread
> > <
> https://github.com/notifications/unsubscribe-auth/AAAPVV663IKYDARR6NBYCU3QM4AUFANCNFSM4IJHKXWQ
> >
> > .
> >
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openzipkin/zipkin/pull/2736?email_source=notifications&email_token=AABQNSEXM2EHHAVV3Q2NU3LQM4IFRA5CNFSM4IJHKXW2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEALFRBQ#issuecomment-538335366>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AABQNSEWI6ECIUUOH7UGAYLQM4IFRANCNFSM4IJHKXWQ>
> .
>

codefromthecrypt commented:
no duration isnt same as 1 duration. I prefer we dont fix like that. there
is a span cleaner javascript

On Fri, Oct 4, 2019, 7:39 PM Anuraag Agrawal <notifications@github.com>
wrote:

> I didn't mean changing the query - just if the JavaScript encounters a span
> with no duration it can fill in 1. The problem that is solved is that
> instead of every call site having to check for empty duration, only one
> does. It's not expensive.
>
> On Fri, Oct 4, 2019, 19:03 Adrian Cole <notifications@github.com> wrote:
>
> > I dont quite understand what problem we are solving here. duration query
> is
> > expensive so we certainly should not add a constraint like that by
> default
> > especially since the query is trace scoped not span scoped. a 1us query
> > will match even a trace that has spans minus a duration.
> >
> >
> > also note that Netflix span here was a converted one from a different
> > system. it is not representative of a proper span, but a flushed span can
> > also be missing a duration.
> >
> > On Fri, Oct 4, 2019, 4:58 PM Anuraag Agrawal <notifications@github.com>
> > wrote:
> >
> > > If the API leaves out duration in some cases, it would be nice to set a
> > > duration of 1 when fetching and duration is missing since that is the
> > > restraint set by the model, and then we don't need to have as many
> checks
> > > at business logic call sites.
> > >
> > > —
> > > You are receiving this because you were mentioned.
> > > Reply to this email directly, view it on GitHub
> > > <
> >
> https://github.com/openzipkin/zipkin/pull/2736?email_source=notifications&email_token=AAAPVV2IUJUJP7R5CP5E2LDQM4AUFA5CNFSM4IJHKXW2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEALAKJY#issuecomment-538314023
> > >,
> > > or mute the thread
> > > <
> >
> https://github.com/notifications/unsubscribe-auth/AAAPVV663IKYDARR6NBYCU3QM4AUFANCNFSM4IJHKXWQ
> > >
> > > .
> > >
> >
> > —
> > You are receiving this because you were mentioned.
> > Reply to this email directly, view it on GitHub
> > <
> https://github.com/openzipkin/zipkin/pull/2736?email_source=notifications&email_token=AABQNSEXM2EHHAVV3Q2NU3LQM4IFRA5CNFSM4IJHKXW2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEALFRBQ#issuecomment-538335366
> >,
> > or mute the thread
> > <
> https://github.com/notifications/unsubscribe-auth/AABQNSEWI6ECIUUOH7UGAYLQM4IFRANCNFSM4IJHKXWQ
> >
> > .
> >
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openzipkin/zipkin/pull/2736?email_source=notifications&email_token=AAAPVVYDSGWERDVSYAOB5DTQM4TQLA5CNFSM4IJHKXW2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEALMDGY#issuecomment-538362267>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAAPVV47VR72SLPZM6FPCMDQM4TQLANCNFSM4IJHKXWQ>
> .
>

anuraaga commented:
Since the yaml you linked indicated minimum duration is 1, I figured you were trying to say no duration is the same as 1 but apparently not.

Anyways going to go ahead and pull out of this PR since this isn't helping.

codefromthecrypt commented:
@anuraaga About the duration 1 issue. trying to solve the programming cleanliness indeed does help. Duration 1 as a solution raised strong antibodies, but I don't think that was your original intent. I wanted to make sure that didn't happen as then we would have to immediately undo it. I'm sorry as I was literally presenting at the time, it was hard for me to give complete context.

I'm most concerned about adding a 1 duration as that will make it hard to see errors or incomplete spans. Also, I'm worried about alignments and things, or people wondering why it took 1us. We have an existing issue somewhere to alert people about broken or incomplete traces, and don't want to make that harder to do. That's why I had strong pushback about coercing absent duration to 1.

* When I mentioned that duration 1 is the only valid one I meant to emphasize we can use this to our advantage. That knowing this zero is the same as absent. If we have numerical calculations, we can use zero, in other words.
* when I said "span cleaner" I literally meant src/zipkin/span-cleaner.js. For example, here we add an empty list for annotations for reasons such as this.

So, to put it all together :) I didn't say explicitly, wasn't online really or anything and it was hard for me to see the contexts of the different angles.. but here's the advice.

It is normal to have span cleaning, knowing duration is 0 is same as absent, then change span-cleaner to backfill zero duration. I think this will solve the programming problem. Adding (getting total duration of a row, which I think this comment was about) is ok.

Main thing if we do this is make sure code that does percentage skips span with zero duration: we don't yet know their duration. Be careful to check, that's all. I don't even remember if we do *any* percentage anymore in the UI, so this point might be a relic of the old one.

jeqo commented:
nit: `REROOT` or `RESET ROOT` instead `RESET REROOT`

codefromthecrypt commented:
+1 to `RESET ROOT` since double-clicking ideally just moves back one root.. the `RESET ROOT` would only be there to go all the way back to root.

llinder commented:
Changes look really nice!  All of the images showing the progression and feedback in the PR is really helpful so thanks to everyone for taking time on those!

codefromthecrypt commented:
@tacigar no pressure, but what's the status on this?

codefromthecrypt commented:
@tacigar if you are swamped, I can try rebasing and addressing remaining review comments. Generally, it is a good idea to keep large change contained and I think this is nearly there. Lemme know.

tacigar commented:
Sorry, I was busy at another work.
I can restart it now.
I'll finish this work as soon as possible.

tacigar committed:
Simplify styling

tacigar committed:
Use React.memo in TracePage

tacigar committed:
Enhance readability

tacigar committed:
Reset reroot -> Reset root

tacigar committed:
Show root span ID

tacigar committed:
Reduce props warning

tacigar committed:
Fix bug

tacigar commented:
> somethings noticed:
> TraceId: part is not completely precise when re-rooting. I think we should display the span ID when navigated here.. even better if the url could.

Thank you for your feedback!
I agree with you.
But I don't want to make this PR bigger anymore, so I don't work on URL for now.
I only show rerooted span ID in page header.
I'll do it after this PR.

tacigar commented:
PTAL 🙇

codefromthecrypt commented:
I have high confidence. just local testing now.

codefromthecrypt commented:
This is really so good. great work @tacigar!
